rfcs/rfc-bld-20230302-1-3p-development.md

```diff
@@ -0,0 +1,215 @@
+                  python3 ${GITHUB_WORKSPACE}/Scripts/extras/pull_and_build_from_git.py $GITHUB_WORKSPACE/package-system/$PACKAGE_NAME
+                  echo $GITHUB_ACCESS_TOKEN | cosign login https://ghcr.io -u $GITHUB_USERNAME --password-stdin
+ cosign upload ghcr.io/o3de/$PACKAGE_NAME:$VERSION ./$PACKAGE_NAME.tar.xz
+```
+
+$~$
+
+## Potential development workflow
+
+### 3P Package Development Phase
+
+1.  Contributors fork/branch the source and GitHub action scripts from the 3p-package-source repo
+2.  During development they can build and package in their "dev" bucket on Github packages using a Github Action build step contained in the fork  
+    
+3.  The build step uses a Github Action to execute a build script (currently the `[build_packages.py](https://github.com/o3de/3p-package-scripts/blob/main/o3de_package_scripts/build_package.py)` script but can be agnostic), which will consume a primary metadata file for each platform  
+    1.  This primary metadata file will be part of contribution prior to the build (currently the [`package_build_host_list_<platform>.json`](https://github.com/o3de/3p-package-source/blob/main/package_build_list_host_windows.json) file), which will define custom steps specific to each package
+    2.  The package source path could include **`PackageInfo.json`**, a metadata file containing package version and licenses, but one can be generated as part of the build
+4.  If the build step succeeds, a separate Github Actions step will create the OCI image from the output, then upload it to Github Packages. A download location and hash of the package will be generated in the build output. 
+5.  To allow a developer to consume their own packages, they would add the Github Packages CDN URL in [cmake/3rdPartyPackages.cmake](https://github.com/o3de/o3de/blob/development/cmake/3rdPartyPackages.cmake#L32) or add to the **`LY_PACKAGE_SERVER_URLS`** environment variable  
+
+$~$
+
+### 3P Package Contribution Phase
+
+1.  When the contribution is ready, a 3P package PR will be created by the contributor, targeting the 3p-package-source development branch  
+    1.  Prechecks via Github Actions will get triggered on PR creation to validate the **`PackageInfo.json`** file for completeness and the license indicated in the file is "blessed" by the project.
+    2.  Recursive dependencies and licenses **should** be called out, but a secondary license check using [Scancode Toolkit](https://github.com/nexB/scancode-toolkit) may be needed to validate
+2.  A maintainer will manually execute the build step in Github Actions once all prechecks are green. An automated comment referencing a codeowners approval group can be added to the PR, which would notify maintainers.  
+    
+3.  If the build fails, the 3P Package PR is kicked back for re-work
+4.  Once the build step is complete, the package is automatically uploaded to Github packages for the O3DE org. The 3P package PR should be merged into 3p-package-source.

```

Comment:
I'm thinking the upload step should be performed after the PR is merged in 3p-package-source instead of after the build step. For example the PR may still need additional changes even after a successful build.

We could instead use a post merge publish step that a maintainer would approve.

Reply:
Agreed, I'm thinking we can provide a publish step for changes pushed to a branch which a contributor can use to test in their fork instead of building it into the PR build step. Either way is fine, just looking for a way to separate the PR workflow from the deployment workflow.

Reply:
Actually, if the contributor forked the repo, the GHA in their fork would follow along and publish the package in their account. This is outlined in the "3P Package Development Phase" section, but I'll make it more clear that it would be building and publishing in their account fork.

Reply:
Actually, if the contributor forked the repo, the GHA in their fork would follow along and publish the package in their account (at no cost them). This is outlined in the "3P Package Development Phase" section, but I'll make it more clear that it would be building and publishing in their account fork.